### ISSUE TITLE
Discussion: Decouple cancel from barrier

### ISSUE THREAD

**Issue by kwannoel** [risingwavelabs/risingwave#13396]
Support cancel of stream job without barrier. 
- Usually when high barrier latency, some stream job can be stuck, but due to high barrier latency, it cannot be cancelled either.
- Another case: When high barrier latency for first barrier, stream job cannot be created, but cannot be cancelled either, until first barrier pass. If we can decouple cancel from barrier, we can avoid this.

Just noting this down first, didn't think of any ideas on how to support it yet.

**Comment by kwannoel** [comment]
We solve part of the problem, where on cancel, we will always remove the table fragment, such that if there's a recovery triggered, the cluster will clean the stream job still.

**Comment by kwannoel** [comment]
IMO this will lead to a lot of friction for PoC users.

Having to trigger recovery is still high effort.

**Comment by kwannoel** [comment]
Offline discussion with @yezizp2012:
- https://risingwave-labs.slack.com/archives/C0340PNL68P/p1701664338832329.

Highlight some parts:
- Q: What about returning immediately e.g. background_cancel / background_drop mv.
   A: The job itself can be causing high latency, we want to immediately drop actors w/o waiting.
       It will could also consume lots of cpu / mem still.
- Remark from @yezizp2012 :
  > The only way to solve this kind case is to force a recovery to clean and rebuild the whole streaming graph.
But in a normal environment, neither DROP nor CANCEL requests are due to the existence of a problematic job in most cases, and I don’t think we should break the design of modifying the streaming graph via barriers just to be compatible with such one. Maybe we can add a force recovery interface for such cases?

In my experience, it is common for users to experience high latency when they're prototyping queries. It's very common to create a query with high join amplification. (Also we should migrate the join matched metrics to the user dashboard perhaps).

Adding a recovery interface could be OK. But users may feel uncomfortable having to trigger a recovery each time barrier latency becomes high. But this is lower effort it seems. So we could support it first.

I would still like to explore the feasibility of decouple drop / cancel from barrier.

**Comment by BugenZhao** [comment]
Also had a discussion with @fuyufjh earlier.

An intuitive approach could be directly dropping the tokio tasks of the involved actors when dropping a streaming job. However, from the view of the upstream actors, the closing of connections or channels during this procedure can be confused with a network failure, in which case the exception should be propagated instead. Thus, it's significant to exploit a mechanism that allows for graceful notification from the downstream to the upstream of the dropping. We are unsure yet if this will be easy to implement.

----

Besides decoupling with the barrier, given that the accomplishment of the last checkpoint for the streaming jobs to be dropped does not matter at all, another approach could be reordering the `Stop` barrier with any other data messages before in the same epoch. The idea could be similar to unaligned checkpoint or barrier stealing, but it may be more challenging to achieve.

**Comment by kwannoel** [comment]
> Thus, it's significant to exploit a mechanism that allows for graceful notification from the downstream to the upstream of the dropping. We are unsure yet if this will be easy to implement.

I think we should introduce a separate notification channel per executor.

Other mechanisms such as altering executor configurations can then attach to that notification channel.

Currently these are also coupled with the message stream.

**Comment by BugenZhao** [comment]
> Other mechanisms such as altering executor configurations can also benefit and attach to that notification channel.
> 
> Currently we can only notify a node via message stream (essentially barrier messages). But in many cases we don't need to synchronize it on barrier.

Definitely +1 for this. 😄 Similar idea previously proposed at: https://github.com/risingwavelabs/risingwave/pull/13166#pullrequestreview-1709460290

**Comment by BugenZhao** [comment]
I think this could be somehow related to https://github.com/risingwavelabs/risingwave/issues/15490.

**Comment by BugenZhao** [comment]
> I think we should introduce a separate notification channel per executor.

I believe we need a detailed design document for this. The concept of "separate notification channel", also known as "control channel" or "local message" previously, has been proposed many times but was always rejected due to implementation complexity or lack of strong motivation. 😄

**Comment by kwannoel** [comment]
> > I think we should introduce a separate notification channel per executor.
> 
> I believe we need a detailed design document for this. The concept of "separate notification channel", also known as "control channel" or "local message" previously, has been proposed many times but was always rejected due to implementation complexity or lack of strong motivation. 😄

Agree: https://github.com/risingwavelabs/rfcs/pull/81#issuecomment-2019522615

**Comment by st1page** [comment]
> > I think we should introduce a separate notification channel per executor.
> 
> I believe we need a detailed design document for this. The concept of "separate notification channel", also known as "control channel" or "local message" previously, has been proposed many times but was always rejected due to implementation complexity or lack of strong motivation. 😄

FYI: 
we used to implement a draft https://github.com/risingwavelabs/risingwave/pull/4834

**Comment by kwannoel** [comment]
Two strong reasons:
1. Let MV/sink/index drop immediately.
2. Let scale complete immediately, this is critical under high barrier latency scenarios, if we want online scaling.

**Comment by kwannoel** [comment]
Discussed with @wenym1 offline some other alternatives:

For CN:
1. Use some global structure in the CNs to contain the state of active / inactive jobs. Inactive jobs are those which are about to be dropped.
2. On drop / cancel, mark the job as inactive. For upstream jobs, if they fail to yield, they can check if the job is inactive using the global structure, and if so, ignore any errors due to this failure.
3. Drop the CN actors.

**Comment by wenym1** [comment]
Following is a general design for this.

##  Things to clean for online drop/cancel job handling:

### meta node
* jobs in the catalog
* meta store table: fragments, actors (deprecating), fragment relations (cleaned cascade on cleaning fragments)
* scheduled command in the queue to global barrier manager, if not processed by global barrier manager yet
* inflight streaming job info in global barrier manager
* revert CreatingStreaming command to None in pending barrier queue if there is any

### compute node
* running actors of the jobs
* entries in dispatcher of any upstream actors
* entries in UpstreamSinkUnion executor for sink-into-table of any downstream actors
* state table ids in hummock uploader
* actor ids in local barrier manager to wait for barrier collection

## Implementation

### Meta node procedure:
1. user issue drop/cancel job ddl.
2. clean jobs in catalog, remove fragments, actors and fragment relations from meta store table.
3. issue to drop/cancel command to `ScheduledBarriers`
4. in `ScheduledBarriers`, we have a separate special queue to handle drop/cancel
  * if the command to create the job is not dequeued yet, remove the command from the queue, and return
  * if the command is not in the command queue, enqueue the cancel/drop command to the special queue to be handled by global barrier manager
5. when global barrier manager receive the cancel/drop command
  * remove the job from inflight streaming job info
  * revert CreatingStreaming command to None in pending barrier queue if there is any
  * (TBD) how to handle if the command is in Completing stage
  * issue a clean request to all CNs to the StreamingControl request stream, save an entry in global barrier manager to store the pending cleaning state in all CNs
  * after all CNs have responded with the clean response, ack to cancel command
6. ack the drop/cancel job rpc request

### How CN local barrier manager handle clean request
#### Cleaning running actors of the jobs
In `Actor`, we pass a oneshot channel `stop_rx` to it. In the outer-most logic near `DispaterExecutor`, we concurrently poll the `stop_rx` while doing other original processing logic. When `stop_rx` is triggered, the actor just gracefully kill itself. The corresponding `stop_tx` can be stored in local barrier manager

#### Cleaning entries in dispatcher of any upstream actors
Downstream actors of the dropping/cancelling job can be killed at anytime, and the dispatcher of upstream actors will fail with error. A simple implementation is that on dispatche error, instead of failing on the error, the actor will simply remove the corresponding entry of the failed target downstream, and don't dispatch to the target anymore. A usual connection error due to network issue can also cause dispatcher error, and when this happen, we expect to trigger a recovery. The upstream actors don't have to handle it either. Such error will be detected by the running downstream merge executor, and report error in the downstream actors to trigger recovery.

#### Cleaning entries in UpstreamSinkUnion executor for sink-into-table of any downstream actors
Extended from the design of cleaning running actors above. When the `DispatcherExecutor` receives `stop_rx`, before it kills itself, it can send a stop message to the dispatch channel, to notify that the upstream is terminating gracefully. When the downstream `UpstreamSinkUnion` detect that the upstream is terminating gracefull, it will just remove the entry of upstream input, and stop polling messages from it.

#### Rest handling
* clean up state table ids of the jobs in hummock uploader
* clean up actor ids in local barrier manager to wait for barrier collection. Should be careful that, there can be inflight BarrierCollectEvent of the actors to drop in the local barrier manager channel
